src/PrettyPrint.hs

```diff
@@ -0,0 +1,224 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module PrettyPrint where
+
+import Syntax
+import qualified Data.Text as T
+
+
+class PrettyPrint a where
+    prettyPrint :: Int -> a -> T.Text
+
+indent :: Int -> T.Text
+indent i = T.concat (replicate i "\t")
+
+instance PrettyPrint (Expr a) where
+    prettyPrint i (Primitive p _ _) = 
+        indent i <> prettyPrint i p
+    prettyPrint i (Var (Identifier var) _ _) = 
+        indent i <> var
+    prettyPrint i (Constructor (ConstructorName var) _ _) = 
+        indent i <> var
+    prettyPrint i (QualifiedVar (ModName mod) (Identifier var) _ _) = 
+        indent i <> mod <> "\\" <> var
+    prettyPrint i (QualifiedConstructor (ModName mod) (ConstructorName c) _ _) =
+        indent i <> mod <> "\\" <> c
+    prettyPrint i (And l r _ _) = 
+        indent i <> prettyPrint 0 l <> " and " <> prettyPrint 0 r
+    prettyPrint i (Or l r _ _) = 
+        indent i <> prettyPrint 0 l <> " or " <> prettyPrint 0 r
+    prettyPrint i (If cond then' else' _ _) = 
+        indent i <> ("if " <> prettyPrint  0 cond) <> "\n" 
+                 <> indent (i + 1) <> "then " <> prettyPrint 0 then' <> "\n"
+                 <> indent (i + 1) <> "else " <> prettyPrint 0 else' <> "\n"
+    prettyPrint i (Block exprs _ _) = 
+        indent i <> "{\n" <> T.concat (map aux exprs) <> indent i <> "}\n"
+      where
+        aux expr = prettyPrint (i + 1) expr <> ";\n"
+    prettyPrint i (LetIn def expr _) = 
+        prettyPrint i def <> " in\n" <> prettyPrint (i + 1) expr
+    prettyPrint i (Lambda pat expr _ _ _) = 
+        indent i <> "fn " 
+                 <> prettyPrint 0 pat <> " => " 
+                 <> prettyPrint 0 expr
+    prettyPrint i (Tuple exprs _ _) = 
+        indent i 
+            <> "(" 
+            <> T.concat (map ((<> ", ") . prettyPrint 0) exprs) 

```

Comment:
I should have written a custom map but this printer was only written to make testing the parser a bit easier so I didn't think that it was worth the time back then.

Reply:
While the problem with block expression isn't of high priority, the one with Tuples and Arrays is quite misleading even for testing.

For now it can even be something like this:
```
mapWithoutLast :: (a -> a) -> [a] -> [a]
mapWithoutLast func []        = []
mapWithoutLast func [x]      = [x]
mapWithoutLast func (x:xs) = f x : mapWithoutLast func xs

```
The fixes may be done later in separate branch or pushed with another feature (TypeChecker with some limitations is "almost" done)